### ISSUE TITLE
questions about boundary conditions

### ISSUE THREAD

**Issue by kmdeck** [CliMA/ClimaCore.jl#83]
hi! not sure if this counts as an "issue", but here are some things that have come up as questions as Ive been prototyping the soil hydrology model...

for background, Richards equation in 1d takes this form: https://en.wikipedia.org/wiki/Richards_equation. Theta is the prognostic variable, and K(theta) and h(theta) are known functions of the state variable appearing on the RHS. The flux is K(theta)[d_z h+1].

In the first example I worked on, the BC are theta_surf = something and grad(h)_bottom = bottom_bc.
https://github.com/CliMA/ClimaCore.jl/blob/b3f34fb08635be0830c07d14a89c6f9943b3ec32/examples/column/richards_vg.jl#L96
If I understand correctly, that means that the value of h_bottom_face (and therefore theta_bottom_face) is whatever is required in order to produce the bottom bc, given the current value of h_bottom_center. I need to interpolate K to be on the faces in order to produce K*grad(h). Shouldnt K_bottom_face be consistent with h_bottom_face (same theta_bottom_face)? 

Is it possible to set a value BC on the flux (-Kgrad(h))? I can't figure out how - it seems like both BC need to be defined in order to get grad(h) at the faces. (Operators.GradientC2F() requires two boundary conditions, and Extrapolation wont work)

I am also curious about moving the boundary conditions out of the RHS function. Right now, for every set of BC, a new RHS function (or method) has to be defined.

**Comment by kmdeck** [comment]
As a concrete example of the 2nd question, in ClimateMachine we could choose a dirichlet condition on the state:
https://github.com/CliMA/ClimateMachine.jl/blob/74964c5f54af2f4d22e9229736f9d353e5ef2c0e/src/Land/Model/soil_bc.jl#L56
and also set the value of the normal flux at a face (which we called Neumann):
https://github.com/CliMA/ClimateMachine.jl/blob/74964c5f54af2f4d22e9229736f9d353e5ef2c0e/src/Land/Model/soil_bc.jl#L124

**Comment by kmdeck** [comment]
hopefully this clarifies the first question:
[Note Aug 2, 2021.pdf](https://github.com/CliMA/ClimaCore.jl/files/6919073/Note.Aug.2.2021.pdf)

**Comment by simonbyrne** [comment]
> Is it possible to set a value BC on the flux (-Kgrad(h))?

Yes, I think you want:

```
gradc2f = Operators.GradientC2F(top = bc_t)
If = Operators.InterpolateC2F(top = bc_ktop)

gradf2c = Operators.GradientF2C(bottom = Operators.SetValue(0.0)) # set the bottom flux to zero

return @. dθl = gradf2c( If(K) * gradc2f(h))
```
As long as all the operators are fused into 1 dotted expression, you shouldn't need to specify `bottom` bcs for `gradc2f` or `If`, since the `gradf2c` bc should intercept the stencil before then.

**Comment by kmdeck** [comment]
as @simonbyrne wrote, the resolution is that as long as we dont compute intermediate quantities, we only need to give two boundary conditions. If we compute intermediate quantities (gradc2f(h), or If(K)), then we do need to give additional BC, and run the risk of not being consistent. This also allows us to apply BC onthe flux itself (-Kgradh) directly.

We also discussed how we can use the parameter object to pass in anything we want!  and values of p can be changed, we think, for e.g. diagnostic variables that update each step. For now, I am using it to pass in boundary condition types that the rhs! function can then dispatch on (e.g. dirichlet, neumann, freedrainage, runoff_bc).
https://github.com/CliMA/ClimaCore.jl/blob/b3444dd9d60c4b6727fd4f7438739869134db20d/examples/column/richards_vg.jl#L33
https://github.com/CliMA/ClimaCore.jl/blob/b3444dd9d60c4b6727fd4f7438739869134db20d/examples/column/richards_vg.jl#L86
https://github.com/CliMA/ClimaCore.jl/blob/b3444dd9d60c4b6727fd4f7438739869134db20d/examples/column/richards_vg.jl#L197

In ClimateMachine, we used Model <: BalanceLaw to store everything (params, IC, BC, etc). Could `p` take that role in general?
